DuckDuckGo/RemoteMessagingStore.swift

```diff
@@ -0,0 +1,354 @@
+                guard let message = remoteMessageManagedObject.message,
+                      let remoteMessage = RemoteMessageMapper.fromString(message),
+                      let id = remoteMessageManagedObject.id
+                else {
+                    continue
+                }
+
+                scheduledRemoteMessage = RemoteMessageModel(id: id, content: remoteMessage.content, matchingRules: [], exclusionRules: [])
+                break
+            }
+        }
+        return scheduledRemoteMessage
+    }
+
+    func fetchRemoteMessage(withId id: String) -> RemoteMessageModel? {
+        var remoteMessage: RemoteMessageModel?
+        context.performAndWait {
+            let fetchRequest: NSFetchRequest<RemoteMessageManagedObject> = RemoteMessageManagedObject.fetchRequest()
+            fetchRequest.predicate = NSPredicate(format: "id == %@", id)
+            fetchRequest.returnsObjectsAsFaults = false
+
+            guard let results = try? context.fetch(fetchRequest) else { return }
+
+            for remoteMessageManagedObject in results {
+                guard let message = remoteMessageManagedObject.message,
+                      let remoteMessageMapped = RemoteMessageMapper.fromString(message),
+                      let id = remoteMessageManagedObject.id
+                else {
+                    continue
+                }
+
+                remoteMessage = RemoteMessageModel(id: id, content: remoteMessageMapped.content, matchingRules: [], exclusionRules: [])
+                break
+            }
+        }
+        return remoteMessage
+    }
+
+    func hasShownRemoteMessage(withId id: String) -> Bool {
+        var shown: Bool = true
+        context.performAndWait {
+            let fetchRequest: NSFetchRequest<RemoteMessageManagedObject> = RemoteMessageManagedObject.fetchRequest()
+            fetchRequest.fetchLimit = 1
+            fetchRequest.predicate = NSPredicate(format: "id == %@", id)
+
+            guard let results = try? context.fetch(fetchRequest) else { return }
+
+            if let result = results.first {
+                shown = result.shown
+            }
+        }
+        return shown
+    }
+
+    func hasDismissedRemoteMessage(withId id: String) -> Bool {
+        var dismissed: Bool = true
+        context.performAndWait {
+            let fetchRequest: NSFetchRequest<RemoteMessageManagedObject> = RemoteMessageManagedObject.fetchRequest()
+            fetchRequest.fetchLimit = 1
+            fetchRequest.predicate = NSPredicate(format: "status == %i", RemoteMessageStatus.dismissed.rawValue)
+
+            guard let results = try? context.fetch(fetchRequest) else { return }
+
+            if results.first != nil {
+                dismissed = true
+            }
+        }
+        return dismissed
+    }
+
+    func dismissRemoteMessage(withId id: String) {
+        updateRemoteMessage(withId: id, toStatus: .dismissed)
+        invalidateRemoteMessagingConfigs()
+    }
```

Comment:
Since this class supports concurrency, shouldn't these kind of multi-step functions be done in transactional way (to ensure no change will "sneak" between these calls)?

Reply:
That's done